File: tools/engine.py
"""
engine.py — Python port of src/game/engine.js

Mirrors every rule exactly. When engine.js changes, update this file in the same PR.
See tools/LEVELGEN.md for the authoritative sync checklist.

Food chain: G(grass) <- R(rabbit) <- F(fox) <- W(wolf) <- B(bear) <- D(dino)
Movement: one step in any of the 8 directions (Chebyshev distance == 1).
Forced-eat priority: lowest trophic tier first (R before F before W before B before D).
  - Exactly 1 forced eat at the active tier  → resolve automatically.
  - 2+ forced eats at the active tier        → player must choose one (forced-choice phase).
Level start constraint: no predator may be 8-dir adjacent to its prey at turn 0.
"""

from __future__ import annotations
import copy
from typing import Optional

# ── Entity tokens ──────────────────────────────────────────────────────────────
GRASS  = 'G'
RABBIT = 'R'
FOX    = 'F'
WOLF   = 'W'
BEAR   = 'B'
DINO   = 'D'
EMPTY  = None

GRID_SIZE = 6

# What each predator eats
PREY_OF: dict[str, str] = {
    RABBIT: GRASS,
    FOX:    RABBIT,
    WOLF:   FOX,
    BEAR:   WOLF,
    DINO:   BEAR,
}

PREY_POINTS: dict[str, int] = {
    GRASS:  1,
    RABBIT: 3,
    FOX:    8,
    WOLF:   20,
    BEAR:   48,
}

# Lower number = resolved first
JUMP_PRIORITY: dict[str, int] = {
    RABBIT: 0,
    FOX:    1,
    WOLF:   2,
    BEAR:   3,
    DINO:   4,
}

DIRS8 = [
    (-1, -1), (-1, 0), (-1, 1),
    ( 0, -1),          ( 0, 1),
    ( 1, -1), ( 1, 0), ( 1, 1),
]

Grid = list[list[Optional[str]]]


# ── Helpers ────────────────────────────────────────────────────────────────────

def clone_grid(g: Grid) -> Grid:
    return [row[:] for row in g]


def in_bounds(r: int, c: int) -> bool:
    return 0 <= r < GRID_SIZE and 0 <= c < GRID_SIZE


def serialize_grid(grid: Grid) -> str:
    return '|'.join(''.join(cell or '.' for cell in row) for row in grid)


def is_one_step(r1: int, c1: int, r2: int, c2: int) -> bool:
    return max(abs(r1 - r2), abs(c1 - c2)) == 1


def get_remaining_score_potential(grid: Grid) -> int:
    return sum(PREY_POINTS.get(cell, 0) for row in grid for cell in row)


def has_illegal_adjacency(grid: Grid) -> bool:
    """True if any predator starts 8-dir adjacent to its prey (forbidden at level start)."""
    for r in range(GRID_SIZE):
        for c in range(GRID_SIZE):
            cell = grid[r][c]
            if not cell or cell == GRASS:
                continue
            prey = PREY_OF.get(cell)
            if not prey:
                continue
            for dr, dc in DIRS8:
                nr, nc = r + dr, c + dc
                if in_bounds(nr, nc) and grid[nr][nc] == prey:
                    return True
    return False


# ── Move logic ─────────────────────────────────────────────────────────────────

def get_legal_targets(grid: Grid, sr: int, sc: int) -> set[tuple[int, int]]:
    piece = grid[sr][sc]
    if not piece or piece == GRASS:
        return set()
    out = set()
    for dr, dc in DIRS8:
        nr, nc = sr + dr, sc + dc
        if not in_bounds(nr, nc):
            continue
        target = grid[nr][nc]
        if target is None or PREY_OF.get(piece) == target:
            out.add((nr, nc))
    return out


def _sort_jump_options(options: list[dict]) -> list[dict]:
    return sorted(
        options,
        key=lambda o: (
            JUMP_PRIORITY[o['pred']],
            o['from'][0], o['from'][1],
            o['to'][0],   o['to'][1],
        )
    )


def get_forced_jump_options(grid: Grid) -> list[dict]:
    """
    Returns every forced-eat at the active (lowest) trophic tier.
    Mirrors getForcedJumpOptions in engine.js exactly.
    """
    options = []
    for r in range(GRID_SIZE):
        for c in range(GRID_SIZE):
            pred = grid[r][c]
            if not pred or pred == GRASS:
                continue
            prey = PREY_OF.get(pred)
            if not prey:
                continue
            for dr, dc in DIRS8:
                nr, nc = r + dr, c + dc
                if in_bounds(nr, nc) and grid[nr][nc] == prey:
                    options.append({
                        'pred': pred,
                        'prey': prey,
                        'from': (r, c),
                        'to':   (nr, nc),
                        'pts':  PREY_POINTS.get(prey, 0),
                        'priority': JUMP_PRIORITY[pred],
                        'kind': 'auto',
                    })

    if not options:
        return []
    min_priority = min(o['priority'] for o in options)
    return _sort_jump_options([o for o in options if o['priority'] == min_priority])


def get_forced_choice(grid: Grid) -> Optional[list[dict]]:
    """Returns forced-choice options when the player must choose, else None."""
    options = get_forced_jump_options(grid)
    return options if len(options) > 1 else None


def _apply_jump_option(grid: Grid, option: dict, kind: str = 'auto') -> dict:
    g = clone_grid(grid)
    fr, fc = option['from']
    tr, tc = option['to']
    earned = PREY_POINTS.get(option['prey'], 0)
    g[tr][tc] = option['pred']
    g[fr][fc] = None
    return {'grid': g, 'pts': earned, 'event': {**option, 'kind': kind}}


def resolve_jumps(grid: Grid) -> dict:
    """
    Cascades auto-eats until the board is quiescent or a forced-choice is required.
    Returns {'grid', 'pts', 'events'}.
    """
    current = clone_grid(grid)
    pts = 0
    events = []
    while True:
        options = get_forced_jump_options(current)
        if len(options) != 1:
            break
        resolved = _apply_jump_option(current, options[0], 'auto')
        current = resolved['grid']
        pts += resolved['pts']
        events.append(resolved['event'])
    return {'grid': current, 'pts': pts, 'events': events}


def _resolve_choice_branches(grid: Grid) -> list[dict]:
    choice = get_forced_choice(grid)
    if not choice:
        return [{'grid': grid, 'pts': 0, 'events': []}]

    outcomes = []
    for option in choice:
        chosen   = _apply_jump_option(grid, option, 'choose')
        after    = resolve_jumps(chosen['grid'])
        tails    = _resolve_choice_branches(after['grid'])
        for tail in tails:
            outcomes.append({
                'grid':   tail['grid'],
                'pts':    chosen['pts'] + after['pts'] + tail['pts'],
                'events': [chosen['event'], *after['events'], *tail['events']],
            })
    return outcomes


def execute_move(grid: Grid, sr: int, sc: int, tr: int, tc: int) -> Optional[dict]:
    """
    Validates and applies one player move, then resolves cascades.
    Returns {'grid', 'pts', 'events'} or None if illegal.
    """
    mover = grid[sr][sc]
    if not mover or mover == GRASS:
        return None
    if not is_one_step(sr, sc, tr, tc):
        return None

    target = grid[tr][tc]
    g = clone_grid(grid)
    pts = 0
    events = []

    if target is not None:
        if PREY_OF.get(mover) != target:
            return None
        earned = PREY_POINTS.get(target, 0)
        pts += earned
        events.append({
            'pred': mover, 'prey': target,
            'from': (sr, sc), 'to': (tr, tc),
            'pts': earned, 'kind': 'hunt',
        })
        g[tr][tc] = mover
        g[sr][sc] = None
    else:
        g[tr][tc] = mover
        g[sr][sc] = None

    auto = resolve_jumps(g)
    pts += auto['pts']
    events.extend(auto['events'])
    return {'grid': auto['grid'], 'pts': pts, 'events': events}


def has_any_move(grid: Grid) -> bool:
    for r in range(GRID_SIZE):
        for c in range(GRID_SIZE):
            p = grid[r][c]
            if not p or p == GRASS:
                continue
            for dr, dc in DIRS8:
                nr, nc = r + dr, c + dc
                if not in_bounds(nr, nc):
                    continue
                t = grid[nr][nc]
                if t is None or PREY_OF.get(p) == t:
                    return True
    return False


# ── Solvability search ─────────────────────────────────────────────────────────

def can_reach_objective(
    grid: Grid,
    objective: dict,
    moves_left: int,
    score: int = 0,
) -> bool:
    """
    DFS with memoisation. Mirrors canReachObjective in engine.js.
    objective: {'type': 'score', 'target': int}
    """
    memo: set[str] = set()

    def dfs(g: Grid, s: int, m: int) -> bool:
        if objective['type'] == 'score' and s >= objective['target']:
            return True
        if m <= 0:
            return False
        if objective['type'] == 'score':
            if s + get_remaining_score_potential(g) < objective['target']:
                return False

        key = f"{serialize_grid(g)}|{m}|{s}"
        if key in memo:
            return False
        memo.add(key)

        for r in range(GRID_SIZE):
            for c in range(GRID_SIZE):
                piece = g[r][c]
                if not piece or piece == GRASS:
                    continue
                for tgt in get_legal_targets(g, r, c):
                    tr, tc = tgt
                    res = execute_move(g, r, c, tr, tc)
                    if not res:
                        continue
                    for branch in _resolve_choice_branches(res['grid']):
                        if dfs(branch['grid'], s + res['pts'] + branch['pts'], m - 1):
                            return True
        return False

    return dfs(grid, score, moves_left)

File: src/game/engine.js
import { GRASS, PREY_OF, PREY_POINTS, DIRS8, GRID_SIZE } from './constants';

export const cloneGrid = (g) => g.map((r) => [...r]);

export const serializeGrid = (grid) =>
  grid.map((row) => row.map((cell) => cell || '.').join('')).join('|');

export const inBounds = (r, c) =>
  r >= 0 && r < GRID_SIZE && c >= 0 && c < GRID_SIZE;

export const cellKey = (r, c) => `${r},${c}`;

export const isOneStep = (r1, c1, r2, c2) =>
  Math.max(Math.abs(r1 - r2), Math.abs(c1 - c2)) === 1;

export const hasIllegalAdjacency = (grid) => {
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      const cell = grid[r][c];
      if (!cell || cell === GRASS) continue;
      const prey = PREY_OF[cell];
      if (!prey) continue;
      for (const [dr, dc] of DIRS8) {
        const nr = r + dr, nc = c + dc;
        if (inBounds(nr, nc) && grid[nr][nc] === prey) return true;
      }
    }
  }
  return false;
};

export const getRemainingScorePotential = (grid) =>
  grid.flat().reduce((sum, cell) => sum + (PREY_POINTS[cell] || 0), 0);

const JUMP_PRIORITY = {
  R: 0,
  F: 1,
  W: 2,
  B: 3,
  D: 4,
};

const sortJumpOptions = (options) =>
  options.sort((a, b) =>
    JUMP_PRIORITY[a.pred] - JUMP_PRIORITY[b.pred] ||
    a.from[0] - b.from[0] ||
    a.from[1] - b.from[1] ||
    a.to[0] - b.to[0] ||
    a.to[1] - b.to[1]
  );

const applyJumpOption = (grid, option, kind = option.kind || 'auto') => {
  const g = cloneGrid(grid);
  const [fr, fc] = option.from;
  const [tr, tc] = option.to;
  const earnedPts = PREY_POINTS[option.prey] || 0;
  g[tr][tc] = option.pred;
  g[fr][fc] = null;
  return {
    grid: g,
    pts: earnedPts,
    event: {
      pred: option.pred,
      prey: option.prey,
      from: option.from,
      to: option.to,
      pts: earnedPts,
      kind,
    },
  };
};

export const getLegalTargets = (grid, sr, sc) => {
  const piece = grid[sr][sc];
  if (!piece || piece === GRASS) return new Set();
  const out = new Set();
  for (const [dr, dc] of DIRS8) {
    const nr = sr + dr, nc = sc + dc;
    if (!inBounds(nr, nc)) continue;
    const target = grid[nr][nc];
    if (target === null || PREY_OF[piece] === target) {
      out.add(cellKey(nr, nc));
    }
  }
  return out;
};

export const getForcedJumpOptions = (grid) => {
  const options = [];

  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      const pred = grid[r][c];
      if (!pred || pred === GRASS) continue;
      const prey = PREY_OF[pred];
      if (!prey) continue;

      for (const [dr, dc] of DIRS8) {
        const nr = r + dr, nc = c + dc;
        if (inBounds(nr, nc) && grid[nr][nc] === prey) {
          options.push({
            pred,
            prey,
            from: [r, c],
            to: [nr, nc],
            pts: PREY_POINTS[prey] || 0,
            priority: JUMP_PRIORITY[pred],
            kind: 'auto',
          });
        }
      }
    }
  }

  if (!options.length) return [];
  const minPriority = Math.min(...options.map((option) => option.priority));
  return sortJumpOptions(options.filter((option) => option.priority === minPriority));
};

export const getForcedChoice = (grid) => {
  const options = getForcedJumpOptions(grid);
  // UI resolves these options as predator-origin -> prey-destination pairs during forced-choice drag.
  return options.length > 1 ? options : null;
};

export const applyForcedChoice = (grid, option) => applyJumpOption(grid, option, 'choose');

export const resolveJumps = (grid) => {
  let current = cloneGrid(grid);
  let pts = 0;
  const events = [];

  while (true) {
    const options = getForcedJumpOptions(current);
    if (options.length !== 1) break;

    const resolved = applyJumpOption(current, options[0], 'auto');
    current = resolved.grid;
    pts += resolved.pts;
    events.push(resolved.event);
  }

  return { grid: current, pts, events };
};

const resolveChoiceBranches = (grid) => {
  const choice = getForcedChoice(grid);
  if (!choice) return [{ grid, pts: 0, events: [] }];

  const outcomes = [];
  for (const option of choice) {
    const chosen = applyForcedChoice(grid, option);
    const afterAuto = resolveJumps(chosen.grid);
    const tails = resolveChoiceBranches(afterAuto.grid);
    for (const tail of tails) {
      outcomes.push({
        grid: tail.grid,
        pts: chosen.pts + afterAuto.pts + tail.pts,
        events: [
          chosen.event,
          ...afterAuto.events,
          ...tail.events,
        ],
      });
    }
  }

  return outcomes;
};

export const canReachObjective = (grid, objective, movesLeft, score = 0) => {
  const memo = new Set();

  const dfs = (currentGrid, currentScore, remainingMoves) => {
    if (objective.type === 'score' && currentScore >= objective.target) return true;
    if (remainingMoves <= 0) return false;

    if (
      objective.type === 'score' &&
      currentScore + getRemainingScorePotential(currentGrid) < objective.target
    ) {
      return false;
    }

    const key = `${serializeGrid(currentGrid)}|${remainingMoves}|${currentScore}`;
    if (memo.has(key)) return false;
    memo.add(key);

    for (let r = 0; r < GRID_SIZE; r++) {
      for (let c = 0; c < GRID_SIZE; c++) {
        const piece = currentGrid[r][c];
        if (!piece || piece === GRASS) continue;
        const targets = getLegalTargets(currentGrid, r, c);
        for (const target of targets) {
          const [tr, tc] = target.split(',').map(Number);
          const res = executeMove(currentGrid, r, c, tr, tc);
          if (!res) continue;

          const branches = resolveChoiceBranches(res.grid);
          for (const branch of branches) {
            if (dfs(branch.grid, currentScore + res.pts + branch.pts, remainingMoves - 1)) {
              return true;
            }
          }
        }
      }
    }

    return false;
  };

  return dfs(grid, score, movesLeft);
};

export const getAmbiguousJump = (grid) => getForcedChoice(grid);

export const executeMove = (grid, sr, sc, tr, tc) => {
  const mover = grid[sr][sc];
  if (!mover || mover === GRASS) return null;
  if (!isOneStep(sr, sc, tr, tc)) return null;

  const target = grid[tr][tc];
  const g = cloneGrid(grid);
  let pts = 0;
  const events = [];

  if (target !== null) {
    if (PREY_OF[mover] !== target) return null;
    const earnedPts = PREY_POINTS[target] || 0;
    pts += earnedPts;
    events.push({
      pred: mover,
      prey: target,
      from: [sr, sc],
      to: [tr, tc],
      pts: earnedPts,
      kind: 'hunt',
    });
    g[tr][tc] = mover;
    g[sr][sc] = null;
  } else {
    g[tr][tc] = mover;
    g[sr][sc] = null;
  }

  const auto = resolveJumps(g);
  pts += auto.pts;
  auto.events.forEach((event) => events.push(event));

  return { grid: auto.grid, pts, events };
};

export const hasAnyMove = (grid) => {
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      const p = grid[r][c];
      if (!p || p === GRASS) continue;
      for (const [dr, dc] of DIRS8) {
        const nr = r + dr, nc = c + dc;
        if (!inBounds(nr, nc)) continue;
        const t = grid[nr][nc];
        if (t === null || PREY_OF[p] === t) return true;
      }
    }
  }
  return false;
};

export const checkWin = (objective, score, maxCombo, grid) => {
  if (objective.type === 'score') return score >= objective.target;
  if (objective.type === 'combo') return maxCombo >= objective.target;
  if (objective.type === 'clear') {
    const apex = findApex(grid);
    return grid.flat().every((c) => !c || c === apex);
  }
  return false;
};

const findApex = (grid) => {
  const order = ['D', 'B', 'W', 'F', 'R', 'G'];
  for (const t of order) {
    if (grid.flat().includes(t)) return t;
  }
  return null;
};
